### ISSUE TITLE
Huge performance impact when TextEdit has large amount line of strings.

### ISSUE THREAD

**Issue by AllenDang** [slint-ui/slint#2306]
I used a readonly TextEdit to display logs.
I noticed that the performance of slint will be hugely impacted when there are more than 1k line of strings are set to text property of TextEdit.
Any hint to improve the performance?

OS: macOS
slint: 0.3.5
backend: winit + femtovg

**Comment by tronical** [comment]
This is a known bug :-(. The text layout itself is not optimised yet for long text blocks and the text edit is missing a model/view design that scales better with large amounts of text blocks.

**Comment by AllenDang** [comment]
@tronical Any walk-around? Maybe ScrollView + a list of Text? Or ListView?

**Comment by ogoffart** [comment]
A ListView with many Text elements might indeed be better since only the visible Text elements will be laid out and rendered.

**Comment by tronical** [comment]
So I've marked a few other issues as duplicate of this, to use this as the collective issue to improve text rendering performance. Now that we've unified the text rendering behind parley, there are two things that we can do fairly easily that should help quite a bit:

1. We split large texts into paragraphs (`\n`) and create `parley::Layout` instances. We could clone the central fontique collection and source cache, send them into `par_iter()` and in there create the layout builder and finally layout in parallel. Shaping is CPU intense, so doing that in threads might be quite helpful for bursts of large documents.

2. We can use the `ItemCache` mechanism that renderers like Skia or FemtoVG use to cache textures to also cache parley layouts in a refcell. That way for example when resizing, the shaping is done once but the line-breaks are done on existing layouts and that's much faster. And rendering also doesn't require re-shaping text (as we currently do!).

I think both are relatively straight-forward to implement, with potential big benefits.

For editing, the ability to modify individual layouts per paragraph would also help with making edits much faster, instead of "re-doing" the entire document when only a paragraph changed.

**Comment by redstrate** [comment]
I noticed while benchmarking large ListViews, that the biggest (and most obvious) hotspots seems to be text layouting - would that be relevant to discuss in this issue as well?

Edit: Digging deeper, it's not simply "text layouting is slow" there's other unrelated issues I think 😬 (like with the animation driver)

**Comment by tronical** [comment]
Splitting text into paragraphs is done, but what remains is the second item:

- Each renderer using parley should also get an `ItemCache` for caching the parley layouts. I think probably most of sharedparley::Layout.
- Each entry in the item cache is connected to a property tracker, so when creating these cached layouts, it's crucial to query all the properties needed to invalidate the cache when they change and **not** query any unrelated properties.
- When text is placed in layouts, it's quite likely that we may want to lay out the same text multiple times to determine the right minimum/maximum sizes. In light of that, it might make sense to cache the text regardless of line breaks and always apply the line breaking (outside of the caching mechanism). That way we should be able to cache at least the layout building and shaping.

**Comment by tronical** [comment]
The caching is implemented now. I'll close this issue. If there's anything else that's actionable for this, then I think we could also track it in #10087
